discovery/watcher.go

```diff
@@ -0,0 +1,339 @@
+	// this to create your gRPC clients. The gRPC connection is
+	// automatically updated to switch to a new server, so you can
+	// use this connection for the lifetime of the associated
+	// Watcher.
+	GRPCConn grpc.ClientConnInterface
+
+	// Token is the ACL token obtain from logging in (if applicable).
+	// If login is not supported this will be set to the static token
+	// from the Config object.
+	Token string
+
+	// Address is the address of current the Consul server the Watcher is using.
+	Address Addr
+
+	// DataplaneFeatures contains the dataplane features supported by the
+	// current Consul server.
+	DataplaneFeatures map[string]bool
+}
+
+// SubscribeState is the info callers need to know when the Watcher's
+// current server changes.
+// TODO: Fill this in.
+type SubscribeState struct{}
+
+type Watcher struct {
+	config Config
+	log    hclog.Logger
+
+	currentServer atomic.Value
+
+	initComplete atomic.Value
+	conn         *grpc.ClientConn
+	resolver     *watcherResolver
+	token        string
+
+	backoff backoff.BackOff
+
+	once sync.Once
+
+	// interface to allow us to inject custom server ports for tests
+	discoverer Discoverer
+}
+
+type serverState struct {
+	addr              Addr
+	dataplaneFeatures map[string]bool
+}
+
+func NewWatcher(config Config, log hclog.Logger) *Watcher {
+	if log == nil {
+		log = hclog.NewNullLogger()
+	}
+
+	// TODO: config for backoff values
+	backoff := backoff.NewExponentialBackOff()
+	backoff.MaxElapsedTime = 0 // Allow backing off forever.
+
+	w := &Watcher{
+		config:     config,
+		log:        log,
+		backoff:    backoff,
+		resolver:   newResolver(log),
+		discoverer: NewNetaddrsDiscoverer(config, log),
+	}
+	w.initComplete.Store(false)
+	w.currentServer.Store(serverState{})
+	return w
+}
+
+func (w *Watcher) Subscribe() chan SubscribeState {
+	// TODO: add this
+	panic("unimplemented")
+}
+
+// Run spawns a goroutine to watch for Consul server set changes.
+// It blocks to wait for initialization to complete, and returns
+// initial state or an error on failure.
+func (w *Watcher) Run(ctx context.Context) (*InitState, error) {
+	if w.conn == nil {
+		var cred credentials.TransportCredentials
+		if tls := w.config.TLS; tls != nil {
+			cred = credentials.NewTLS(tls)
+		} else {
+			cred = insecure.NewCredentials()
+		}
+
+		dialOpts := []grpc.DialOption{
+			grpc.WithTransportCredentials(cred),
+			grpc.WithResolvers(w.resolver), // note: experimental api.
+			// TODO: Add interceptors here.
+			// TODO: Add custom grpc balancer here.
+		}
+
+		// Dial with "consul://" to trigger our custom resolver. We don't
+		// provide a server address. The connection will be updated by the
+		// Watcher via the custom resolver once an address is known.
+		conn, err := grpc.DialContext(ctx, "consul://", dialOpts...)
+		if err != nil {
+			return nil, err
+		}
+		w.conn = conn
+	}
+
+	// Spawn our goroutine.
+	w.once.Do(func() { go w.run(ctx) })
```

Comment:
Though I don't always stick to it, I know many folks prefer to keep the `go` statement in the calling-code (especially with a library like this) to make the goroutine lifecycle clear.

I appreciate in this case it'd mean you couldn't return the `InitState`, though. So I wondered what you thought about separating these methods into `Run` (which starts the run-loop) and `State` (which blocks until initialization and returns the current state).

Reply:
That works for me.

I'm not sure where we should return the error that they currently might get from Run()? I can maybe move the initialization of `w.conn` into the constructor, so that we have `NewWatcher() (*Watcher, error)`. I'll see if this works.

Reply:
Yeah, so I updated things a bit to support this. Mainly:

1. `NewWatcher(context.Context, ...) (*Watcher, error)` - I moved the gRPC connection init stuff to the constructor, so that we can return an error there instead of from the `Run` method.
2. `NewWatcher` also accepts a context since the `grpc.DialContext` needs the context, so I removed the context function parameters and switched to using an internal context (`Watcher.ctx` and `Watcher.ctxCancel`).
3. Added the `State` method like you described. I like that this can return the current state at any time.
4. Added a `Stop` method which cancels the `Run` and waits for it to stop. The internal context from (2) makes this easy.

I like this better! What do you think?